Gh-398: Where to run gaffer administration guide (#406)
gchq/gaffer-doc: docs/administration-guide/gaffer-deployment/gaffer-docker/how-to-run.md

# How to Deploy with Accumulo on Docker

After reading the [previous page](./gaffer-images.md) you should have a good
understanding of what images are available for Gaffer and how to configure them

> to you needs. However, before running a deployment backed by Accumulo you will

Reviewer: ```suggestion
to your needs. However, before running a deployment backed by Accumulo you will
```

need to know a bit of background on Hadoop to understand how the data will scale
and be distributed.

Usually when deploying a container image you simply run the image and everything
is contained locally to the container (hence the name). For larger scale graphs

> this less desireable as we will usually want to be able to scale and load

Reviewer: ```suggestion
this is less desirable as we will usually want to be able to scale and load
```

Accumulo to provide distributed synchronization so is useful to start up first.

```bash
docker run \
       --detach \
       --name zookeeper \
       --hostname zookeeper \
       --net gaffer-example \
       --env ZOO_SERVERS="server.1=zookeeper:2888:3888;2181" \
       --env ZOO_4LW_COMMANDS_WHITELIST="*" \
       --volume /data \
       --volume /datalog \
       zookeeper:3.7.1
```

The above docker command will run a ZooKeeper container with few bits of
configuration. The main part that is being set up is the hostname and ports.
With ZooKeeper you can also do this by providing a `zoo.cfg` file but as there
is not much to do for a small cluster so we can just pass the configuration in
via environment variables with `--env`. See the [official ZooKeeper docs](https://zookeeper.apache.org/)
for more information and configuration options.

### Hadoop/HDFS

Next we can launch the Hadoop cluster, we will use the custom distribution of
the HDFS image `gchq/hdfs`. As mentioned before this provides a single-node
Hadoop cluster which we can run multiple times to extend into a multi-node
cluster.

To run a Hadoop cluster we first need the configuration files for Hadoop which

> we can then add into the running containers. As a start point you can use the

Reviewer: ```suggestion
we can then add into the running containers. As a starting point you can use the
```

           --hostname accumulo-monitor \
           --net gaffer-example \
           --publish 9995:9995 \
           --env ACCUMULO_CONF_DIR="/etc/accumulo/conf" \
           --env HADOOP_USER_NAME="hadoop" \
           --volume /custom/configs/accumulo:/etc/accumulo/conf \
           --volume /var/log/accumulo \
           gchq/gaffer:2.0.0-accumulo-2.0.1 monitor
    ```

=== "tserver node"

    ```bash
    docker run \
           --detach \
           --name accumulo-tserver \
           --hostname accumulo-tserver \
           --net gaffer-example \
           --env ACCUMULO_CONF_DIR="/etc/accumulo/conf" \
           --env HADOOP_USER_NAME="hadoop" \
           --volume /custom/configs/accumulo:/etc/accumulo/conf \
           --volume /var/log/accumulo \
           gchq/gaffer:2.0.0-accumulo-2.0.1 tserver
    ```

!!! note
    Please see the [official Accumulo docs](https://accumulo.apache.org/docs/2.x/configuration/overview)
    for more information on configuring the deployment.

### REST API

The final container we need to start up is the REST API, this essentially gives
the front end so we can use containers together in a Gaffer cluster. The REST

> API container is also where the configuration for the graph is applied such as,

Reviewer: ```suggestion
API container is also where the configuration for the graph is applied, such as
```

the schema files and store properties.

To start up the REST API it is a similar process to the other containers;

> however, there is a few more bind-mounts that need defining to configure the

Reviewer: ```suggestion
however, there are a few more bind-mounts that need defining to configure the
```

# How to Deploy with Accumulo on Docker

After reading the [previous page](./gaffer-images.md) you should have a good
understanding of what images are available for Gaffer and how to configure them
to you needs. However, before running a deployment backed by Accumulo you will
need to know a bit of background on Hadoop to understand how the data will scale
and be distributed.

Usually when deploying a container image you simply run the image and everything
is contained locally to the container (hence the name). For larger scale graphs
this less desireable as we will usually want to be able to scale and load
balance the storage based on the volume of data; this is where Hadoop comes in.

> ## What is Hadoop/Accumulo?

Reviewer 2: This is a nice explanation of Hadoop, but I'm not sure it should live on this page.

Author: The idea was you really should know about it before doing a deployment I didn't think it was big enough to warrant its own page

Reviewer 2: Maybe it could go on the Accumulo Store page?

Author: Moved it to Store page, added link here to it too

[Apache Hadoop](https://hadoop.apache.org/) is an open-source software framework
used for distributed storage and processing of large datasets. Hadoop is
designed to handle various types of data, including structured, semi-structured,
and unstructured data. It is a highly scalable framework that allows users to
add nodes to the cluster as needed.

Hadoop has two main components: Hadoop Distributed File System (HDFS) and
MapReduce. HDFS is a distributed file system that provides high-throughput
access to data. MapReduce is a programming model used for processing large
datasets in parallel.

Accumulo is built on top of the HDFS to provide a key-value store with all the
same scalability and robustness of Hadoop.

> ## Running a Cluster via Docker for Gaffer

Reviewer 2: This is manually walking through the steps which occur when running the docker compose file. I think it would be good to make that clear and link to that file.

Author: Agreed probably should add a link in, yeah the idea was that using standard docker means should be translatable to other tools other than docker compose
